src/main/java/com/mongodb/spark/sql/connector/read/MongoMicroBatchPartitionReader.java

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.mongodb.spark.sql.connector.read;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.read.PartitionReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDocument;
+import org.bson.Document;
+
+import com.mongodb.client.ChangeStreamIterable;
+import com.mongodb.client.MongoChangeStreamCursor;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.model.Aggregates;
+import com.mongodb.client.model.Filters;
+
+import com.mongodb.spark.sql.connector.assertions.Assertions;
+import com.mongodb.spark.sql.connector.config.ReadConfig;
+import com.mongodb.spark.sql.connector.exceptions.MongoSparkException;
+import com.mongodb.spark.sql.connector.schema.BsonDocumentToRowConverter;
+
+/**
+ * A partition reader returned by {@link
+ * MongoPartitionReaderFactory#createReader(org.apache.spark.sql.connector.read.InputPartition)}.
+ * It's responsible for outputting data for a RDD partition.
+ */
+public class MongoMicroBatchPartitionReader implements PartitionReader<InternalRow> {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(MongoMicroBatchPartitionReader.class);
+  private static final String FULL_DOCUMENT = "fullDocument";
+  private final MongoMicroBatchInputPartition partition;
+  private final BsonDocumentToRowConverter bsonDocumentToRowConverter;
+  private final ReadConfig readConfig;
+  private boolean closed = false;
+  private MongoClient mongoClient;
+  private MongoChangeStreamCursor<BsonDocument> changeStreamCursor;
+  private InternalRow currentRow;
+
+  /**
+   * Construct a new instance
+   *
+   * @param partition the partition
+   * @param bsonDocumentToRowConverter the converter from {@link BsonDocument} to {@link
+   *     InternalRow}
+   * @param readConfig the read configuration for reading from the partition
+   */
+  public MongoMicroBatchPartitionReader(
+      final MongoMicroBatchInputPartition partition,
+      final BsonDocumentToRowConverter bsonDocumentToRowConverter,
+      final ReadConfig readConfig) {
+
+    this.partition = partition;
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+    this.readConfig = readConfig;
+    LOGGER.info(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  /** Proceed to next record, returns false if there is no more records. */
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    MongoChangeStreamCursor<BsonDocument> cursor = getCursor();
+    BsonDocument next = cursor.tryNext();
```

Comment:
It's not correct to call `tryNext` here, I don't think. If `tryNext` returns null, it doesn't mean that it won't return non-null in the future.  I think what we need is to:

1. Remove the `resumeToken` filter from the pipeline
2. Replace `tryNext` with `next`
3. Return false from this method if `next.clusterTime` >= `endOffsetTimestamp`.

Reply:
> we are looking in the past so there should always data available now

But the server doesn't guarantee that it will send that data in the current batch.  It reserves the right to send empty batches at any time.  One realistic example: a write with operation time `x` where `x  < endOffsetTimestamp` but doesn't get majority committed until `y` where y` > endOffsetTimestamp`.

Reply:
> What I want to do avoid is any unnecessary blocking
 
I don't think we can avoid blocking given the micro-batch iterator-style API (it doesn't have a tryNext equivalent).

Reply:
I think `tryNext` semantics could be used correctly here.  

* `tryNext()` will call `getMore()` as required. 
* It will return `null` if, the cursor no longer exists, the limit is reached or if there is no results from `getMore`.

We set no limit, so we just care about two scenarios. Cursor no longer exists (i.e closed) and no results available.

We can confirm that no results are ever expected by checking the postBatchResumeToken to ensure that if its timestamp is >= endOffsetTimestamp then there will be no future results.

That would stop any forever blocking which `next()` does if there are no matching results.

Reply:
I think cursor no longer exists results in an error returned from getMore, and is pretty rare (e.g. the collection has been dropped).  No results available is the tricky one.  As I said above, there is no guarantee from the server (afaik at least) that any getMore command response will return a result.  Empty batches are allowed, and are no guarantee that the next batch won't be empty.  So we can use tryNext here, but a loop would be required.  And isn't that effectively the same loop that the driver already has in its implementation on `next`?

Reply:
This has been rewritten and should be clearer in intent now.